Services/UI/OJS.Services.Ui.Business/Validation/Implementations/ContestValidationService.cs

```diff
@@ -0,0 +1,118 @@
+        }
+
+        if (await this.IsContestExpired(contest.Id, userId, isUserAdmin, official))
+        {
+            return ValidationResult.Invalid(ValidationMessages.Contest.IsExpired, ContestValidation.ContestIsNotExpired.ToString());
+        }
+
+        if (official &&
+            !await this.CanUserCompeteByContestByUserAndIsAdmin(
+                contest,
+                userId,
+                isUserAdmin,
+                allowToAdminAlways: true))
+        {
+            return ValidationResult.Invalid(ValidationMessages.Contest.CanBeCompeted, ContestValidation.ContestCanBeCompeted.ToString());
+        }
+
+        if (!official && !contest.CanBePracticed && !isUserLecturerInContest)
+        {
+            return ValidationResult.Invalid(ValidationMessages.Contest.CanBePracticed, ContestValidation.ContestCanBePracticed.ToString());
+        }
+
+        return ValidationResult.Valid();
+    }
+
+    private bool IsUserLecturerInContest(Contest contest, string userId) =>
+        contest.LecturersInContests.Any(c => c.LecturerId == userId) ||
+        contest.Category.LecturersInContestCategories.Any(cl => cl.LecturerId == userId);
+
+    private async Task<bool> IsContestExpired(int contestId, string userId, bool isAdmin, bool official)
+    {
+        var contest = await this.contestsData.GetByIdWithParticipants(contestId);
+        if (contest == null)
+        {
+            return false;
+        }
+
+        var isUserAdminOrLecturerInContest = isAdmin || await this.contestsData
+            .IsUserLecturerInByContestAndUser(contestId, userId);
+
+        if (contest.IsOnline && !isUserAdminOrLecturerInContest)
+        {
+            var participant = contest.Participants.FirstOrDefault(p => p.UserId == userId && p.IsOfficial);
+
+            if (participant == null)
+            {
+                if (!official && contest.PracticeEndTime.HasValue)
+                {
+                    return DateTime.Now > contest.PracticeEndTime;
+                }
+
+                if (official && contest.EndTime.HasValue)
+                {
+                    return DateTime.Now > contest.EndTime;
+                }
+
+                return false;
+            }
+
+            return participant.ParticipationEndTime >= DateTime.Now;
+        }
+
+        return false;
+    }
+
+    private async Task<bool> CanUserCompeteByContestByUserAndIsAdmin(
+        Contest contest,
+        string userId,
+        bool isAdmin,
+        bool allowToAdminAlways = false)
+    {
+        var isUserAdminOrLecturerInContest = isAdmin || await this.contestsData
+            .IsUserLecturerInByContestAndUser(contest.Id, userId);
+
+        if (contest.CanBeCompeted || (isUserAdminOrLecturerInContest && allowToAdminAlways))
+        {
+            return true;
+        }
+
+        if (isUserAdminOrLecturerInContest && contest.IsActive)
+        {
+            return true;
+        }
+
+        return false;
```

Comment:
Extracting the `if` logic into 2 methods, this method will become more expressive and understandable:
```
var isUserAdminOrLecturerInContest = isAdmin || await this.contestsData
            .IsUserLecturerInByContestAndUser(contest.Id, userId);
return this.IsAccessibleToLecturerOrAdmin(...) || this.CanUserCompete(...);
```

Reply:
Done.